cms/lib/xblock/authoring_mixin.py

```diff
@@ -43,3 +44,10 @@ def visibility_view(self, _context=None):
         fragment.add_javascript_url(self._get_studio_resource_url('/js/xblock/authoring.js'))
         fragment.initialize_js('VisibilityEditorInit')
         return fragment
+
+    copied_from_block = Reference(
+        # Note: used by the content_staging app. This field is not needed in the LMS.
```

Comment:
FWIW, since this kind of connective tissue is useful for analytics, I'm not sure if folks might want it emitted there in the LMS as part of the event data. @bmtcril?

Reply:
@ormsbee Where do you think this would be emitted on the LMS side? Right now we're pulling xblock data from the CMS, I'm not sure why we would want it on the LMS side of the house. Maybe in cases where authoring is happening outside of CMS?

Reply:
I didn't know if we wanted this to be emitted as a field in the tracking log events. If you don't think that's an immediate need, that's fine. We're not locked into it, since we can move it to somewhere that'll be accessible from both later if that comes up.

Reply:
@ormsbee @Agrendalath I just found another issue.

If I now try to copy a block that has been previously pasted, it gives an error because of the new `copied_from_block` field:

```
  File "/openedx/edx-platform/openedx/core/lib/xblock_serializer/block_serializer.py", line 63, in _serialize_normal_block
    block.add_xml_to_node(olx_node)
  ...
  File "/openedx/venv/lib/python3.8/site-packages/xblock/mixins.py", line 489, in add_xml_to_node
    self._add_field(node, field_name, field)
  File "/openedx/venv/lib/python3.8/site-packages/xblock/mixins.py", line 526, in _add_field
    value = field.to_string(field.read_from(self))
  File "/openedx/venv/lib/python3.8/site-packages/xblock/fields.py", line 620, in to_string
    value = json.dumps(
  ...
TypeError: Object of type BlockUsageLocator is not JSON serializable
```

It seems that `Reference` field types don't serialize properly using the XBlock vanilla API, and there is special logic in edx-platform to handle them and make them aware of OpaqueKeys, but it doesn't integrate with the XBlock APIs that we're using here (`add_xml_to_node`, `parse_xml` etc.), only with the [mongo modulestore](https://github.com/openedx/edx-platform/blob/f09f3543c498ff5236d519d27ca44f438f427f96/xmodule/modulestore/mongo/base.py#L1436-L1459) and the [whole-course xml exporter/importer](https://github.com/openedx/edx-platform/blob/f09f3543c498ff5236d519d27ca44f438f427f96/xmodule/modulestore/xml_exporter.py#L353-L377).

My question is: can/should I just use a `String` field instead to store this "original usage key", or should I find a (probably very hacky) way to get it to work as a `Reference` ?

It also seems like there is code that will sometimes rewrite `Reference` field values [to override the `course` part of the key](https://github.com/openedx/edx-platform/blob/f09f3543c498ff5236d519d27ca44f438f427f96/xmodule/modulestore/xml_exporter.py#L363) which we definitely don't want in general, but [in other places it says](https://github.com/openedx/edx-platform/blob/f09f3543c498ff5236d519d27ca44f438f427f96/xmodule/modulestore/xml_importer.py#L803-L810) it will only do that "if the original namespace matched the original course" which does sound like desirable behavior. Overall though exact effect using of `Reference` fields seems unclear to me.

Though I guess I'll have to fix this "properly" somehow eventually anyways, or XBlocks that use `Reference`/`ReferenceList`/`ReferenceValueDict` won't be copy-pastable? Actually in the core platform I can only see `ConditionalBlock` using any of these fields, and that's it. Maybe just using a string is better then if these `Reference` fields aren't really used in practice anyways?

Reply:
For now I have just converted it to a string, and the problem is gone.

Reply:
Honestly, I had forgotten that `Reference` even exists. String does everything we need in this case, so I'm all for using it instead.